<h1 style="text-align: center;">Linear Regression Analysis</h1>
<h1 style="text-align: center;">Homework 4</h1>
<h3 style="text-align: center;">Brandon Miner</h3>
<h3 style="text-align: center;">September 25, 2025</h3>

**Directions**: Submit a .pdf file containing your responses. The .pdf can be converted from a Latex file, pictures of your handwritten solutions, word files, markdown files, etc. If there are coding problems, upload a separate notebook for Python code.

### *Written Questions*

1. Suppose we have fit a MLR model between response variable $Y$ and predictors $X_1,\dots,X_{p−1}$. using a data of size $n$. The global F-test aka omnibus test considers the hypotheses:
$$
H_0 : y_i = \beta_0 + \epsilon_i \text{ vs. } 
H_1 : y_i = \beta_0 | \beta_1X_{1i} + \dots + \beta_{p-1}X_{(p-1)i}+\epsilon_i
$$
using the statistic $F = \frac{MSR}{MSE}$ where $MSR$ and $MSE$ are calculated under the full model. Show that this definition is equivalent to the alternative formulation of the F statistic as:

$$
F_{alt} = 
\frac{
    \frac{SSE_{H_0} − SSE_{H_1}}
        {df_{SSE_{H_0}} −df_{SSE_{H_1}}}}
    {\frac{SSE_{H_1}}
        {df_{SSE_{H_1}}}}
$$


**ANSWER**
$$
F_{alt} = 
\frac{
    \frac{SSE_{H_0} − SSE_{H_1}}
        {df_{SSE_{H_0}} −df_{SSE_{H_1}}}}
    {\frac{SSE_{H_1}}
        {df_{SSE_{H_1}}}}
=
\frac{
    \frac{SSE_{H_0} − SSE_{H_1}}
        {n − (n-1)}}
    {\frac{SSE_{H_1}}
        {n-2}}
=
\frac{
    SSE_{H_0} − SSE_{H_1}}
    {\frac{SSE_{H_1}}
        {n-2}}
$$


from here realize that
$$
MSE = \frac{SSE_{H_1}}{n-2},\space MSR= SSE_{H_0} - SSE_{H_1}
$$
Because this is the definition of MSE, and MSR is the remaining error of the reduced model.

Thus
$$
= \frac{MSR}{MSE}
$$
As desired

2. If a predictor variable is categorical with six states and we want to include it in a regression model, how many dummy variables do we need to use?

**ANSWER** 

You would include 5 dummy variables because having the same number of dummy variables as categories induces perfect multicollinearity. Mathematically, this makes our predictors $X$ invertible, and thus $X^TX$ invertible.

3. Suppose a predictor variable is categorical with three states “C1”, “C2”, “C3”. When we include it in a regression model and the individual t tests used “C1” as reference level, and showed “C2” is significant and “C3” is not. Would you conclude that “we should drop C3 and fit a new model”? Why or why not?

**ANSWER**

I would not immediately drop C3 because it could be possible that C3 is simply correlated with C1 or C2, and the true model could be $Y=\beta \text{C3}$. Further exploration is needed to conclude *why* C3 is not significant.

### *Coding Questions*

4. This question will help you to understand the calculation of ANOVA in MLR using an example. For
the dataset `KelleyBlueBookData.csv`, response= Price against the following predictors: Mileage, Liter, Cylinder (in this order). Treat Cylinder as a quantitative variable.

In [23]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

kbbd = pd.read_csv("../../data/KelleyBlueBookData.csv")

model_4 = ols('Price~Mileage+Liter+Cylinder',data=kbbd).fit()
model_4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.342
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     138.8
Date:                Thu, 25 Sep 2025   Prob (F-statistic):           2.18e-72
Time:                        18:31:55   Log-Likelihood:                -8367.7
No. Observations:                 804   AIC:                         1.674e+04
Df Residuals:                     800   BIC:                         1.676e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4707.6150   1602.866      2.937      0.003    1561.296    7853.934
Mileage       -0.1544      0.035     -4.461      0.000      -0.222      -0.086
Liter       1545.2522    893.411      1.730      0.084    -208.454    3298.958
Cylinder    2847.9345    712.040      4.000      0.000    1450.247    4245.622
==============================================================================
Omnibus:                      214.158   Durbin-Watson:                   0.074
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              444.825
Skew:                           1.499   Prob(JB):                     2.56e-97
Kurtosis:                       5.071   Cond. No.                     1.37e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.37e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

(a) Run the sequential ANOVA for the fitted model. Report null and alternative hypothesis, the F
stat, and the p-value for the F-test for dropping or including the ‘Cylinder’ predictor. What is
the conclusion of this test?

In [33]:
model_anova = sm.stats.anova_lm(model_4,typ=1)
model_anova

df        sum_sq       mean_sq           F        PR(>F)
Mileage     1.0  1.605590e+09  1.605590e+09   24.890171  7.448191e-07
Liter       1.0  2.421824e+10  2.421824e+10  375.435819  7.134623e-69
Cylinder    1.0  1.031948e+09  1.031948e+09   15.997457  6.930502e-05
Residual  800.0  5.160560e+10  6.450701e+07         NaN           NaN

**ANSWER**

The Null Hypothesis is that Cylinder is not a significant predictor of price. 

The Alternative Hypothesis is that Cylinder is a significant predictor of price.

The F-stat is 15.997457.

The p-value is 6.930502e-05

Putting this all together, we can reject the null hypothesis at a 5% significance level. Cylinder appears to be a significant predictor of price. 

(b) Manually run the test in part (a) yourself: 1. fit the null model in python and extract $SSE$ and
degrees of freedom of this $SSE$; then 2. fit the alternative model in python and extract $SSE$ and
degrees of freedom of this $SSE$. Plug in the numbers to

$$
F_{alt} = 
\frac{
    \frac{SSE_{H_0} − SSE_{H_1}}
        {df_{SSE_{H_0}} −df_{SSE_{H_1}}}}
    {\frac{SSE_{H_1}}
        {df_{SSE_{H_1}}}}
$$

Does the value you calculated match the F-statistic from part (a)?

In [48]:
F_stat = model_anova.loc["Cylinder","F"]

null_model = ols('Price~Mileage+Liter',data=kbbd).fit()
null_anova = sm.stats.anova_lm(null_model)

SSEH0 = null_anova.loc['Residual','sum_sq']
dfSSEH0 = null_anova.loc['Residual','df']

SSEH1 = model_anova.loc['Residual','sum_sq']
dfSSEH1 = model_anova.loc['Residual','df']

Falt = ((SSEH0 - SSEH1) / (dfSSEH0 - dfSSEH1)) / (SSEH1/dfSSEH1)

print(f"F-stat: {F_stat}\nF-alt-stat: {Falt}\n\nDifference: {F_stat - Falt}")


F-stat: 15.99745707625833
F-alt-stat: 15.99745707625828

Difference: 4.973799150320701e-14


**ANSWER**

We know from Q1 that the F stat and F-alt stat should be the same. Here we can see that they are effectively the same, only off due to minor rounding error.

(c) Run the partial ANOVA (typ=2) for the fitted model. Does the F-test for ’Cylinder’ match the
F-test from part (a)? Why or why not?

In [54]:
model_anova_2 = sm.stats.anova_lm(model_4,typ=2)
F_stat_2 = model_anova_2.loc["Cylinder","F"]

print(f"seq F-stat: {F_stat}\nMin-One F-stat: {F_stat_2}")

print(f" \nDifference: {F_stat - F_stat_2}")

seq F-stat: 15.99745707625833
Min-One F-stat: 15.997457076258485
 
Difference: -1.545430450278218e-13


**ANSWER**

The F-tests are very similar. This makes sense if we assume that the number of cylinders are not correlated with the other predictors and a strong predictor of price. This would result in both tests passing strongly.

(d) From the partial ANOVA (typ=2) table in (c), report the null and alternative hypothesis, the F
stat, and the p-value for the F-test for dropping or including the ‘Mileage’ predictor. Interpret
the result of this test.

In [56]:
model_anova_2.loc['Cylinder','PR(>F)']

np.float64(6.930501986611917e-05)

**ANSWER**

The Null Hypothesis is that Cylinder is not a significant predictor of price. 

The Alternative Hypothesis is that Cylinder is a significant predictor of price.

The F-stat is 15.997457076258485

The p-value is 6.930501986611917e-05

Putting this all together, we can reject the null hypothesis at a 5% significance level. Cylinder appears to be a significant predictor of price. 

(e) Manually run the test in part (d) yourself: 1. fit the null model in python and extract SSE and
degrees of freedom of this SSE; then 2. fit the alternative model in python and extract SSE and
degrees of freedom of this SSE. Plug in the numbers to

$$
F_{alt} = 
\frac{
    \frac{SSE_{H_0} − SSE_{H_1}}
        {df_{SSE_{H_0}} −df_{SSE_{H_1}}}}
    {\frac{SSE_{H_1}}
        {df_{SSE_{H_1}}}}
$$

Does the value you calculated match the F-statistic from part (d)?

In [58]:
F_stat_2 = model_anova_2.loc["Cylinder","F"]

null_model = ols('Price~Mileage+Liter',data=kbbd).fit()
null_anova = sm.stats.anova_lm(null_model, typ=2)

SSEH0 = null_anova.loc['Residual','sum_sq']
dfSSEH0 = null_anova.loc['Residual','df']

SSEH1 = model_anova.loc['Residual','sum_sq']
dfSSEH1 = model_anova.loc['Residual','df']

Falt = ((SSEH0 - SSEH1) / (dfSSEH0 - dfSSEH1)) / (SSEH1/dfSSEH1)

print(f"F-stat: {F_stat}\nF-alt-stat: {Falt}\n\nDifference: {F_stat - Falt}")


F-stat: 15.997457076258485
F-alt-stat: 15.99745707625828

Difference: 2.042810365310288e-13


**ANSWER**

They are similar which makes sense mathematically as talked about prior as well as logically since the different F_stats for typ1 and typ2 are the same.

5. This question will help you to understand the calculation of $R^2$ and $R^2_{adj}$ in MLR using an example. For the dataset `KelleyBlueBookData.csv`:

(a) Fit Model 1: a model which considers Price as the response and regresses it against the predictors Mileage and Cylinder. Report the $R^2$ and $R^2_{adj}$ values from the summary table.

In [59]:
model_5a = ols("Price~Mileage+Cylinder",data=kbbd).fit()
model_5a.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.340
Model:                            OLS   Adj. R-squared:                  0.338
Method:                 Least Squares   F-statistic:                     206.2
Date:                Thu, 25 Sep 2025   Prob (F-statistic):           5.95e-73
Time:                        19:04:49   Log-Likelihood:                -8369.2
No. Observations:                 804   AIC:                         1.674e+04
Df Residuals:                     801   BIC:                         1.676e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   3145.7503   1325.934      2.372      0.018     543.034    5748.467
Mileage       -0.1524      0.035     -4.401      0.000      -0.220      -0.084
Cylinder    4027.6746    204.612     19.684      0.000    3626.036    4429.313
==============================================================================
Omnibus:                      198.944   Durbin-Watson:                   0.077
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              385.493
Skew:                           1.439   Prob(JB):                     1.96e-84
Kurtosis:                       4.797   Cond. No.                     1.01e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.01e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**ANSWER**

R-squared:	0.340

Adj. R-squared:	0.338

Calculate $R^2$ and $R^2_{adj}$ for the model in part (a) yourself. Obtain the SSE and SST of the model in part (a), then plug in the formulas: 
$R^2 = 1 − \frac{SSE}{SST}$ and 
$R^2_{adj} = 1 - \frac{SSE/n −p}{SST/n −1}$.
Do the values match with the python output in (a)

In [64]:
import numpy as np


SSE = np.sum((kbbd['Price'] - model_5a.fittedvalues) ** 2)
SST = np.sum((kbbd['Price'] - np.mean(kbbd['Price'])) ** 2)
n = len(kbbd)
R_squared = 1 - SSE/SST
R_adj_squared = 1 - (SSE / (n-2)) / (SST / n-1)

print(f"R-Squared : {R_squared}\nR-adj-Squared : {R_adj_squared}")

R-Squared : 0.33982070826263056
R-adj-Squared : 0.3381743690825033


**ANSWER**

Asside from some rounding error, the values are effectively the same.

(c) Fit Model 2: a model which considers Price as the response and regresses it against the predictors Mileage, Liter and Cylinder. Report the $R^2$ and $R^2_{adj}$ values from the summary table. Which
model is preferable according to R2adj between Model 1 and Model 2? Why?

In [65]:
model_5a = ols("Price~Mileage+Liter+Cylinder",data=kbbd).fit()
model_5a.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.342
Model:                            OLS   Adj. R-squared:                  0.340
Method:                 Least Squares   F-statistic:                     138.8
Date:                Thu, 25 Sep 2025   Prob (F-statistic):           2.18e-72
Time:                        19:20:07   Log-Likelihood:                -8367.7
No. Observations:                 804   AIC:                         1.674e+04
Df Residuals:                     800   BIC:                         1.676e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4707.6150   1602.866      2.937      0.003    1561.296    7853.934
Mileage       -0.1544      0.035     -4.461      0.000      -0.222      -0.086
Liter       1545.2522    893.411      1.730      0.084    -208.454    3298.958
Cylinder    2847.9345    712.040      4.000      0.000    1450.247    4245.622
==============================================================================
Omnibus:                      214.158   Durbin-Watson:                   0.074
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              444.825
Skew:                           1.499   Prob(JB):                     2.56e-97
Kurtosis:                       5.071   Cond. No.                     1.37e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.37e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**ANSWER**

R-squared:	0.342

Adj. R-squared:	0.340

According to Adj. R-squared, I'd prefer model 1 because it has a larger Adj. R-squared. But it is important to note how small both Adj. R-squared values they are.

Open question: Consider simultaneously the t-test results, ANOVA, $R^2_{adj}$ and any other concepts
we have covered so far (e.g. diagnostics). Which model would you choose, Model 1 or Model 2?
Argue for your model in terms of these statistics and also the real life meaning of the problem.

**ANSWER**

I would use Model 2 because, while its Adj. R-squared is lower, it's not that much lower and liter seems to be a the best predictor of price out of iter, mileage, and cylinder based on its incredibly small p-value from the sequential ANOVA. This leads me to believe the true model contains liter.